### PR TITLE
HPCC-12658 Enable C++11 and fix various compile warnings

### PR DESCRIPTION
**Pull request by ghalliday**
See also https://github.com/hpcc-systems/LN/pull/758 in the LN repository

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman please have a look

File: thorlcr/graph/thgraph.hpp [hunk 1]
```diff
@@ -1030,7 +1030,7 @@ class graph_decl CActivityBase : public CInterface, implements IExceptionHandler
 
 // IExceptionHandler
     bool fireException(IException *e);
-    void processAndThrowOwnedException(IException * e) __attribute__((noreturn));
+    __declspec(noreturn) void processAndThrowOwnedException(IException * e) __attribute__((noreturn));
```
Comment (richardkchapman) [inline thorlcr/graph/thgraph.hpp]: Is this the only instance of this? Seems unlikely

File: tools/esdlcomp/CMakeLists.txt [hunk 2]
```diff
@@ -47,7 +47,11 @@ include_directories (
          ${CMAKE_CURRENT_SOURCE_DIR}
     )
 
-ADD_DEFINITIONS( -D_CONSOLE -DESDLCOMP_EXPORTS -O0 )
+ADD_DEFINITIONS( -D_CONSOLE -DESDLCOMP_EXPORTS )
+if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_CLANG)
+    ADD_DEFINITIONS( -O0 )
```
Comment (richardkchapman) [inline tools/esdlcomp/CMakeLists.txt]: Why are we using -O0 at all?

**Comment by richardkchapman** [comment]
I slightly wonder whether we should separate out the warning fixes from the enabling of C++11 - separate commits at least, just in case one causes issues.

File: thorlcr/graph/thgraph.hpp (hunk 1 above)
Comment (ghalliday) [inline thorlcr/graph/thgraph.hpp]: Most definitions have both the windows and linux annotations - this only had linux.

File: tools/esdlcomp/CMakeLists.txt (hunk 2 above)
Comment (ghalliday) [inline tools/esdlcomp/CMakeLists.txt]: No idea.  Possibly it takes too long to compile the state tables.

**Comment by ghalliday** [comment]
See #7855 and #7854 instead.

**Pull request closed without merging**